src/xenia/patcher/patcher.cc

```diff
@@ -0,0 +1,217 @@
+        XELOGE("Unsupported patch allocation size");
+        break;
+    }
+    // Restore previous protection
+    heap->Protect(patch_data_entry.memory_address_,
+                  patch_data_entry.alloc_size_, old_address_protect);
+
+    if (!is_any_patch_applied_) {
+      is_any_patch_applied_ = true;
+    }
+  }
+}
+
+PatchFileEntry PatchingSystem::ReadPatchFile(
+    const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
+  std::shared_ptr<cpptoml::table> patch_toml_fields;
+
+  try {
+    patch_toml_fields = cpptoml::parse_file(file_path.u8string());
+  } catch (...) {
+    XELOGE("Cannot load patch file: {}", file_path.u8string());
+    patchFile.title_id = -1;
+    return patchFile;
+  };
+
+  auto title_name = patch_toml_fields->get_as<std::string>("title_name");
+  auto title_id = patch_toml_fields->get_as<std::string>("title_id");
+  auto title_hash = patch_toml_fields->get_as<std::string>("hash");
+
+  patchFile.title_id = strtoul((*title_id).c_str(), NULL, 16);
+  patchFile.hash = strtoull((*title_hash).c_str(), NULL, 16);
+  patchFile.title_name = *title_name;
+
+  auto tarr = patch_toml_fields->get_table_array("patch");
+
+  for (auto table : *tarr) {
+    PatchInfoEntry patch = PatchInfoEntry();
+    auto patch_name = *table->get_as<std::string>("name");
+    auto patch_desc = *table->get_as<std::string>("desc");
+    auto patch_author = *table->get_as<std::string>("author");
+    auto is_enabled = *table->get_as<bool>("is_enabled");
+
+    patch.id = 0;  // ToDo: Implement id for future GUI stuff
+    patch.patch_name = patch_name;
+    patch.patch_desc = patch_desc;
+    patch.patch_author = patch_author;
+    patch.is_enabled = is_enabled;
+
+    const std::string data_types[] = {"be64", "be32", "be16", "be8"};
+
+    // Iterate through all available data sizes
+    for (const std::string& type : data_types) {
+      auto entries = ReadPatchData(type, table);
+
+      for (const PatchDataEntry& entry : *entries) {
+        patch.patch_data.push_back(entry);
+      }
+    }
+    patchFile.patch_info.push_back(patch);
+  }
+  return patchFile;
+}
+
+std::vector<PatchDataEntry>* PatchingSystem::ReadPatchData(
+    const std::string size_type,
+    const std::shared_ptr<cpptoml::table>& patch_table) {
+  std::vector<PatchDataEntry>* patch_data = new std::vector<PatchDataEntry>();
+  auto patch_data_tarr = patch_table->get_table_array(size_type);
+
+  if (!patch_data_tarr) {
+    return patch_data;
+  }
+
+  for (const auto& patch_data_table : *patch_data_tarr) {
+    auto address = patch_data_table->get_as<std::uint32_t>("address");
+    // Todo: How to handle different sizes
+    auto value = patch_data_table->get_as<std::uint64_t>("value");
+
+    PatchDataEntry patchData = {GetAllocSize(size_type), *address, *value};
+    patch_data->push_back(patchData);
+  }
+  return patch_data;
+}
+
+uint8_t PatchingSystem::GetAllocSize(const std::string provided_size) {
+  uint8_t alloc_size = sizeof(uint32_t);
```

Comment:
Needs better error handling, defaulting to `uint32_t` if an unknown size string is passed in is probably a bad idea.